instrument/invariant.go

```diff
@@ -0,0 +1,44 @@
+// Copyright (c) 2018 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package instrument
+
+const (
+	// InvariantViolatedMetricName is the name of the metric emitted upon
+	// invocation of `EmitInvariantViolation`.
+	InvariantViolatedMetricName = "invariant-violated"
+
+	// InvariantViolatedLogPrefix is a common prefix that should be used
+	// when generating errors/log statements pertaining to the violation
+	// of an invariant.
+	InvariantViolatedLogPrefix = "[invariant violated]"
+)
+
+// EmitInvariantViolation emits a metric to indicate a system invariant has
+// been violated. Users of this method are expected to monitor/alert off this
+// metric to ensure they're notified when such an event occurs. Further, they
+// should log further information to aid diagnostics of the system invariant
+// violated at the callsite of the violation.
+func EmitInvariantViolation(opts Options) {
```

Comment:
This all looks fine to me, my only question is do you think there are any places in the code-base where you might have access to a scope but not the InstrumentOptions? Probably unlikely right?

Reply:
AFAIK all exported entry points into packages across our codebase pass around scopes using InstrumentOptions. Within a package we sometimes pass only the scope around but that's a minor refactor I don't mind expecting people to do if they want to use this method. Motivation is mainly because it feels weird to have a utility method on a scope live here, case Scope doesn't live here. At least if it operates on the InstrumentOptions, it kinda makes sense to keep here.